src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java

```diff
@@ -256,43 +269,56 @@ private void fileAclOp(AclOp op, Dataset dataset, List<String> fileIds, Map<IamR
             }
         }
 
-        Map<String, GoogleBucketResource> bucketCache = new HashMap<>();
+        Map<String, GoogleBucketResource> bucketCache = new ConcurrentHashMap<>();
 
         int batchSize = configurationService.getParameterValue(FIRESTORE_SNAPSHOT_BATCH_SIZE);
         List<List<String>> batches = ListUtils.partition(fileIds, batchSize);
         logger.info("operation {} on {} file ids, in {} batches of {}",
             op.name(), fileIds.size(), batches.size(), batchSize);
 
+        String retrieveTimer = performanceLogger.timerStart();
         int count = 0;
         for (List<String> batch : batches) {
             logger.info("operation {} batch {}", op.name(), count);
             count++;
 
             List<FSFile> files = fileDao.batchRetrieveById(dataset, batch, 0, true);
-            for (FSFile file : files) {
-                // Cache the bucket resources to avoid repeated database lookups
-                GoogleBucketResource bucketForFile = bucketCache.get(file.getBucketResourceId());
-                if (bucketForFile == null) {
-                    bucketForFile = dataLocationService.lookupBucket(file.getBucketResourceId());
-                    bucketCache.put(file.getBucketResourceId(), bucketForFile);
-                }
-                Storage storage = storageForBucket(bucketForFile);
-                URI gsUri = URI.create(file.getGspath());
-                String bucketPath = StringUtils.removeStart(gsUri.getPath(), "/");
-                BlobId blobId = BlobId.of(bucketForFile.getName(), bucketPath);
-                switch (op) {
-                    case ACL_OP_CREATE:
-                        for (Acl acl : acls) {
-                            storage.createAcl(blobId, acl);
-                        }
-                        break;
-                    case ACL_OP_DELETE:
-                        for (Acl.Group group : groups) {
-                            storage.deleteAcl(blobId, group);
-                        }
-                        break;
-                }
+
+            try (Stream<FSFile> stream = files.stream()) {
+                List<Future<FSFile>> futures = stream.map(file -> executor.submit(() -> {
+                    // Cache the bucket resources to avoid repeated database lookups
```

Comment:
I think this would be more readable if you pulled out the unnamed lambda into a method.

Reply:
Fair enough.